workflow/airflow_libero/dags/ingest_article_xml_dag.py

```diff
@@ -0,0 +1,125 @@
+
+DIR_PATH = 'data'
+
+
+def _download_article_xml(*args, **kwargs):
+    """Download some article xml and store in a public location
+
+    :return: str
+    """
+
+    input_data = kwargs['dag_run'].conf.get('input_data')
+
+    if input_data:
+        data_dir = input_data.get('data_dir')
+        file_urls = input_data.get('urls')
+
+        if not os.path.isdir(data_dir):
+            os.mkdir(data_dir)
+
+        if file_urls:
+            for url in file_urls:
+                response = requests.get(url=url)
+
+                file_name = url.split('/')[-1]
+
+                with open(os.path.join(data_dir, file_name), 'w') as xml_file:
+                    xml_file.write(response.text)
+
+
+def _extract_asset_uris(*args, **kwargs):
+    """Download some article xml and extract asset URIs from XML.
+
+    :return: str
+    """
+
+    input_data = kwargs['dag_run'].conf.get('input_data')
+
+    if input_data:
+        file_urls = input_data.get('urls')
+
+        uris = []
+
+        if file_urls:
+            for url in file_urls:
+                response = requests.get(url=url)
+                xml = BeautifulSoup(response.text, 'lxml-xml')
+                uris += [var.contents[0] for var in xml.find_all('source')
+                         if var.attrs['media-type'] == 'image/tiff']
+
+        print(uris)
+
+        kwargs['ti'].xcom_push('uris', json.dumps({'uris': uris}))
+
+
+def _download_assets(*args, **kwargs):
+    """Download some static assets.
+
+    :return: str
+    """
+
+    input_data = kwargs['dag_run'].conf.get('input_data')
+    uris = kwargs['ti'].xcom_pull(task_ids=None, key='uris')
+
+    if input_data and uris:
+        data_dir = input_data.get('data_dir')
+
+        uris = json.loads(uris.replace("'", '"'))
+
+        for uri in uris['uris']:
+            response = requests.get(url=uri)
+
+            file_name = uri.split('/')[-1]
+
+            with open(os.path.join(data_dir, file_name), 'wb') as asset_file:
+                asset_file.write(response.content)
+
+
+default_args = {
+    'owner': 'airflow',
+    'depends_on_past': False,
+    'start_date': airflow.utils.dates.days_ago(2),
+    'email': ['airflow@example.com'],
+    'email_on_failure': False,
+    'email_on_retry': False,
+    'retries': 1,
+    'retry_delay': timedelta(seconds=10)
+}
+
+
+dag = DAG('ingest_article_xml', default_args=default_args, schedule_interval=None)
+
+
+download_article_xml = EventEmittingPythonOperator(task_id='download_article_xml',
+                                                   provide_context=True,
+                                                   python_callable=_download_article_xml,
+                                                   dag=dag)
+
+extract_asset_uris = EventEmittingPythonOperator(task_id='extract_asset_uris',
+                                                 provide_context=True,
+                                                 python_callable=_extract_asset_uris,
```

Comment:
Passing in an object's method rather than a function would be identical right?
Thinking about objects that take collaborators in their constructors, and can then be used as activities.

Reply:
Yes as long as the module you are initiating these operators is the one pulling in and setting up the collab classes.

I don't particularly like how you have to define the dags, but they are just classes at the end of the day if you had a lot of custom dags and tasks/operators then a more abstract library structure would be easier to manage and share.